icelake/src/io/mod.rs

```diff
@@ -17,22 +17,212 @@ pub use scan::*;
 use arrow_array::RecordBatch;
 use arrow_schema::SchemaRef;
 
-use crate::types::DataFileBuilder;
+use crate::types::{DataFileBuilder, StructValue};
 use crate::Result;
+pub mod input_wrapper;
+pub use input_wrapper::*;
+
+type DefaultInput = RecordBatch;
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriter: Send + 'static {
-    async fn write(&mut self, batch: RecordBatch) -> Result<()>;
-    async fn close(&mut self) -> Result<Vec<DataFileBuilder>>;
+pub trait FileWriterBuilder: Send + Sync + Clone + 'static {
+    type R: FileWriter;
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
 }
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriterBuilder: Send + Sync + Clone + 'static {
-    type R: RecordBatchWriter;
+pub trait FileWriter: Send + 'static {
+    type R: FileWriteResult;
+    async fn write(&mut self, batch: &RecordBatch) -> Result<()>;
+    async fn close(self) -> Result<Self::R>;
+}
+
+pub trait FileWriteResult: Send + 'static {
+    type R: IcebergWriteResult;
+    fn empty() -> Self;
+    fn combine(&mut self, other: Self);
+    fn to_iceberg_result(self) -> Self::R;
+}
+
+#[async_trait::async_trait]
+pub trait IcebergWriterBuilder: Send + Clone + 'static {
+    type R;
     async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
 }
 
-pub trait SingletonWriter: RecordBatchWriter {
-    fn current_file(&self) -> String;
+#[async_trait::async_trait]
+pub trait IcebergWriter<I = DefaultInput>: Send + 'static {
+    type R: IcebergWriteResult;
+    async fn write(&mut self, input: I) -> Result<()>;
+    async fn flush(&mut self) -> Result<Self::R>;
```

Comment:
Rename this to `close`? Though it's not safe in api, the name `close` is still more clear that we should not call it after calling.

Reply:
For now the semantic of the flush is different with close: flush all state in this writer and return the result, and it's acceptable that we can call it again. 🤔

Reply:
Hmm, that's weird, could you have some example for this case? I was think for example in risingwave, we can have `IcebergWriterBuilder` stored sink writer, and for every checkpoint, we call `close` method to flush all states, and get the write results.

Reply:
Sounds reasonable.